### PR TITLE
Epub nozip

### PR DESCRIPTION
**Pull request by TylerZeroMaster**
fixes openstax/ce#2019

## Summary
Making just epub jobs support multiple links would have been about as much work as updating all job types. Consequently, it seemed better to do it all.

There is a new file, `$IO_BOOK/slugs`, that may or may not exist. If it exists, only book slugs contained in it are uploaded. If it does not exist, all book slugs in the repository are uploaded. For the PDF job type, this list is also used to select which book slugs to build in the `step-pdf` step. This would be possible in the docx job type as well; however, I thought that would be better as a separate PR (this one is too big already).

- Adds step-upload -epub, -docx, and -pdf to upload one or more artifacts
- Adds `upload_book_artifacts` function used in the new step-upload -epub, -docx, and -pdf to upload one or more artifacts from the build using aws cli
- Refactors step-upload-book so it generates preview urls
- Removes generate_preview_urls script (combined into step-upload-book)
- Adds upload stubbing via environment variable `STUB_UPLOAD`
    - Supports both "corgi" and "webhosting" modes (determines which variables are set with test values)
- Replaces `pdf_url` with `artifact_urls.json`
- Adds testing for all upload steps
    - Tests that the expected arguments were used when calling `aws` and `copy-resources-s3`
    - Tests that the generated `artifact_urls.json` file matches expected output
- Updates tests for all job types to include building with and without book slug argument
- Updates concourse resource
    - Simplified logic for loading artifact urls
    - Added support for arrays of artifact urls
- Removes old variables like `collection_id` and `content_server` from `INPUT_SOURCE_DIR` parsing.

~~Right now the tests fail because of the epub cover error. I will pull that change in when it is ready. If you want to run the tests locally before that change is pulled in, you can add this into `step-epub` with the other errors that are ignored:~~
```bash
grep -v 'Non-linear content must be reachable, but found no hyperlink to "cover.xhtml".'
```

### CODE REVIEW

File: build-concourse/script/override_common_log.sh [hunk 1]
```diff
@@ -1,2 +1,5 @@
 #!/bin/bash
+
+# TODO: Maybe delete this script? It's not currently used.
```
Comment (philschatz) [inline build-concourse/script/override_common_log.sh:3]: :+1:

File: dockerfiles/docker-entrypoint.sh [hunk 2]
```diff
@@ -58,12 +58,49 @@ LOCAL_ATTIC_DIR=${LOCAL_ATTIC_DIR:-}
...
 JS_DEBUG=${JS_DEBUG:-}
 NODE_OPTIONS=${NODE_OPTIONS:-}
+STUB_UPLOAD=${STUB_UPLOAD:-}
 
 if [[ $JS_DEBUG ]]; then
     export NODE_DEBUG_PORT=9229 # LCOV_EXCL_LINE
     export NODE_OPTIONS="--inspect-brk=0.0.0.0:$NODE_DEBUG_PORT" # LCOV_EXCL_LINE
 fi
 
+if [[ $STUB_UPLOAD ]]; then
```
Comment (philschatz) [inline dockerfiles/docker-entrypoint.sh:68]: ooh nice!

File: corgi-concourse-resource/tests/test_resource.py [hunk 3]
```diff
@@ -167,34 +167,51 @@ def test_resource_files_are_written(self):
             job_json[key] = None
         assert job_json == expected_json
 
-        collection_id = read_file(os.path.join(dest_path, "collection_id"))
-        assert collection_id == read_file(os.path.join(DATA_DIR, "collection_id"))
+        repository = read_file(os.path.join(dest_path, "repo"))
+        assert repository == read_file(os.path.join(DATA_DIR, "repo"))
+
+        slugs = read_file(os.path.join(dest_path, "slugs"))
```
Comment (philschatz) [inline corgi-concourse-resource/tests/test_resource.py:173]: :tada: So nice to get away from the `collection_id` file and have repo and slugs in their own file!

**Review by philschatz — APPROVED** [review]
Looks great! Replacing & cleaning up the now-dead references to collection_id & collection_style, stubbing out all the upload code so it can "run" locally, moving the upload bits into tiny similar files, and updating the tests to match are really nice!

**Comment by omehes** [comment]
verified in corgi staging - 20230814.162815
OK

**Comment by omehes** [comment]
deployed and verified in corgi prod - OK

**Pull request merged**